### PR TITLE
SAM E70: fix Ethernet driver cache coherency issues and always enable data cache

### LINKED ISSUE #9812: drivers: eth: gmac: Fix incorrect cache coherency handling code
#8282 added cache support for Atmel SAM GMAC Ethernet driver. Unfortunately the implemented cache coherency management method ensuring coherency of descriptor lists is incorrect.

Atmel SAM Ethernet module is using a scatter-gather technique to exchange data with the Ethernet driver. To communicate the location of the data buffers the driver sets-up a so called descriptor list. This is effectively a place in RAM containing a sequence of 32-bit words representing buffer location and its status. If cache is enabled every time the device driver updates data in the descriptor list it should theoretically also call cache clean operation. This would ensure that the data is written to RAM where it can be read by the hardware module. Unfortunately, at the same time the device driver is modifying one location in the descriptor list the hardware module can be modifying one next to it. Since cache clean operation works on a full cache line - 32 bytes in case of Atmel device - it would overwrite any such modifications done by the hardware module. As such, ensuring cache coherency using standard techniques (cache clean / cache invalidate) is not possible here. 

In case of GMAC driver the cache coherency for the descriptor lists should be implemented using the so called non-cacheable RAM region.

This is described in detail in "Cache Coherence Management Limits" chapter of [How to manage Cortex-M7 Cache Coherence on the Atmel SAM S70 / E70](https://community.atmel.com/sites/default/files/forum_attachments/AN-15679_SAMS70_E70_Cache_Coherence.pdf) Application Note.

### PR DESCRIPTION
**Pull request by aurel32**
This pull request fixes the Atmel SAM Ethernet driver cache coherency issues by using the recently added nocache memory and always enable the data cache. After this the `gmac_desc_*` functions introduced to manage the cache coherency through cache operations look a bit empty. I wonder if we should get rid of them. The same way given the cache is now always enabled, I wonder if we should still check if it is enabled. Anyway this can probably be done in a further cleanup PR.

Fixes #9812

### CODE REVIEW

**Review by tbursztyka — APPROVED** [review]

**Review by tgorochowik — CHANGES_REQUESTED** [review]
Nice! That looks and works just great!

Yeah, the helpers are not really needed that much now, and checking if cache is enabled may be redundant, but I guess it is ok to clean it up later.

Could we at least remove this comment from the header?
https://github.com/zephyrproject-rtos/zephyr/blob/fa5b8b4b506f056c86be9a9a216852dbbbaa66c4/drivers/ethernet/eth_sam_gmac.c#L19

**Comment by aurel32** [comment]
@tgorochowik Thanks for the review and thanks for pointing at that comment. I have just pushed a new version which removes it.

**Review by tgorochowik — APPROVED** [review]
Thanks!

File: drivers/ethernet/eth_sam_gmac.c [hunk 1]
```diff
@@ -152,27 +150,25 @@ static inline void dcache_clean(u32_t addr, u32_t size)
...
-	dcache_invalidate((u32_t)desc, sizeof(struct gmac_desc));
 	return desc->w1;
 }
 
 /* Set operations */
 static inline void gmac_desc_set_w0(struct gmac_desc *desc, u32_t value)
 {
 	desc->w0 = value;
-	dcache_clean((u32_t)desc, sizeof(struct gmac_desc));
+	__DSB();
```
Comment (mnkp) [inline drivers/ethernet/eth_sam_gmac.c]: I believe this line is redundant. In a few places in the driver code where we do care about the order of the memory accesses there is the DMB instruction. So DSB is no longer necessary. Unless there is another issue at hand I'm not aware of.

**Review by jukkar — APPROVED** [review]
Tested this and it works ok. Also tried with a version that removed `__DSB()` calls as suggested by @mnkp , and did not see any issues there either.

**Comment by jukkar** [comment]
Hmm, just after giving my comment, I started to see gptp sync issues. I am using sam-e70 board for gptp testing. I will do some more testing with this patch and a version that removes the __DSB() calls.

**Comment by jukkar** [comment]
The gptp drops from sync eventually regardless of this patch or not, so the issues I am seeing are related to some other problem. +1 from me to this fix.

**Comment by aurel32** [comment]
@mnkp indeed, it looks like the necessary barriers are there, so i'll drop the __DSB() from the next version. Now the `gmac_desc_*` functions are really empty...

I also have found another issue looking at the code, `dcache_clean` is called from a second place in addition to the net buffers: 

https://github.com/zephyrproject-rtos/zephyr/blob/724d6f79092e8a2639563f38e97e71f1389eb39b/drivers/ethernet/eth_sam_gmac.c#L1170-L1174

This is done on a 32-bit value, so less than a cache line, and therefore looks worrying. This has been added in commit a313e5c74f0 by @tgorochowik . I don't really get why as from what I understand the  RX buffer accounting list is not a list shared with Ethernet device. I guess we have two choices there, either drop that cache operation or declare `rx_frag_list_que0/1/2` as nocache.

**Comment by aurel32** [comment]
I ran some additional tests, and it some cases the Ethernet device get stuck for some time, or indefinitely, probably something similar to #11255.

I have found it happens when the Zephyr network stack has to sent 2 packets in row, for example an ack for the just received data + answer data. This seems to be a latent issue, but it happens more often given that instruction cache and now data cache have been enabled. Basically the CPU is now faster than Ethernet device (which is good in some sense).

The problem seems to be the following one:
1. the first packet goes through `eth_tx`, one `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is submitted.
2. the second packet also goes through `eth_tx`, another `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is not started because there is already one already submitted. Not sure that's a problem in that case.
3. the first packet has been sent, one `tx_desc_sem` semaphore is given and the `tx_timeout_work` delayed work is cancelled.
4. the second packet has been sent but given the delayed work has already been cancelled, `tx_completed` is not called: the `tx_desc_sem` semaphore is not given back and the network packet is not unreferenced.

I am not fully sure about how exactly the steps 2 and 3 are interleaved, what is sure is that the call to `tx_completed` is skipped for the second packet has the job has already been cancelled. Adding some delay between sending packets (for instance at the beginning of `eth_tx`) workarounds the issue.

It looks like we should not allow sending a new packet until the previous packet has been fully sent.

**Comment by mnkp** [comment]
@aurel32 Good catch. `dcache_clean` on line 1172 is indeed not needed. Just as you mentioned "the RX buffer accounting list is not a list shared with Ethernet device".

I'll have a look at the other bug you described.

**Comment by aurel32** [comment]
@mnkp, thanks for the confirmation, i'll add a patch removing it.

For the bug, here's the workaround I am using:

```
diff --git a/drivers/ethernet/eth_sam_gmac.c b/drivers/ethernet/eth_sam_gmac.c
index e0e2d4a883..751dfc66ef 100644
--- a/drivers/ethernet/eth_sam_gmac.c
+++ b/drivers/ethernet/eth_sam_gmac.c
@@ -1329,6 +1327,11 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
         */
        net_buf_push(pkt->frags, net_pkt_ll_reserve(pkt));
 
+       /* Wait until the previous packet has been fully sent */
+       while (k_sem_count_get(&queue->tx_desc_sem) != (queue->tx_desc_list.len - 1)) {
+               k_yield();
+       }

```

**Comment by aurel32** [comment]
Note, I am able to reproduce the issue using `samples/net/sockets/dumb_http_server` and running a loop of 100 wget fetching data from it.

**Comment by mnkp** [comment]
> 2\. the second packet also goes through `eth_tx`, another `tx_desc_sem` semaphore is taken, and the `tx_timeout_work` delayed work is not started because there is already one already submitted. Not sure that's a problem in that case.

Well spotted. The fact that the `k_delayed_work_submit()` is not called for the second frame is indeed the problem. And as you notice

> 4\. the second packet has been sent but given the delayed work has already been cancelled, `tx_completed` is not called:

The whole timeout concept has to be reworked. We shouldn't wait until the tx frame is fully sent. In any case the bug is orthogonal to this PR.

File: drivers/ethernet/eth_sam_gmac.c [hunk 2]
```diff
@@ -515,8 +467,7 @@ static void tx_completed(Gmac *gmac, struct gmac_queue *queue)
 			     queue_list[queue->que_idx]);
 #endif
 
-	__ASSERT(gmac_desc_get_w1(&tx_desc_list->buf[tx_desc_list->tail])
-		 & GMAC_TXW1_USED,
+	__ASSERT(&tx_desc_list->buf[tx_desc_list->tail]->w1 & GMAC_TXW1_USED,
```
Comment (mnkp) [inline drivers/ethernet/eth_sam_gmac.c]: `tx_desc_list->buf[tx_desc_list->tail].w1`

**Review by mnkp — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_sam_gmac.c (hunk 2 above)
Comment (aurel32) [reply drivers/ethernet/eth_sam_gmac.c]: Good catch, that's fixed now.

**Comment by aurel32** [comment]
> The whole timeout concept has to be reworked. We shouldn't wait until the tx frame is fully sent. In any case the bug is orthogonal to this PR.

Not fully orthogonal. Enabling the cache means the code sending packets runs faster, triggering the  issue. Maybe I should drop the last commit enabling the cache from the PR?

**Review by mnkp — APPROVED** [review]

**Comment by mnkp** [comment]
> Not fully orthogonal.

True. I guess one of the maintainers should decide if we should fix the timeout bug first or merge partially this PR.

As for the timeout bug - I'm not sure if we should discuss it further here or maybe move the conversation to a separate issue - one possible and easy fix would be to get rid of all the k_delayed_* functions and associated data structures and inside eth_tx where we have
```
		k_sem_take(&queue->tx_desc_sem, K_FOREVER);
```
replace `K_FOREVER` with the timeout value. It works a bit differently than the old design but does the job. We can only get stuck on the tx_queue if there is a bug in the driver or the hardware malfunctions. So this is a safety feature, not part of the normal functional behavior. We would only have to be careful writing the small piece of code that cleans up the data in case we do fail to get the semaphore on time.

**Comment by aurel32** [comment]
This PR is ready, but should go after the fix for #11255, hence the DNM label.

**Comment by aurel32** [comment]
PR #11255 has now been fixed, but this PR now also depends on PR #12548. The DNM label should therefore stay.

**Comment by aurel32** [comment]
This issue has been ready and approved for almost two months, but is blocked by another PR. As the freeze is almost there, I believe it make sense to merge it anyway. I have therefore removed the last commit enabling the cache, I'll do a separate PR for it that can be merged once PR #12817 gets in.

**Pull request merged**